checker/src/org/checkerframework/checker/index/upperbound/UpperBoundStore.java

```diff
@@ -0,0 +1,337 @@
+        for (Element e : enclosedElts) {
+            AnnotatedTypeMirror atm = analysis.getTypeFactory().getAnnotatedType(e);
+            enclosedTypes.add(atm);
+            if (e.getKind() == ElementKind.METHOD) {
+                ExecutableElement ee = (ExecutableElement) e;
+                List<? extends Element> rgparam = ee.getParameters();
+                for (Element param : rgparam) {
+                    AnnotatedTypeMirror atmP = analysis.getTypeFactory().getAnnotatedType(param);
+                    enclosedTypes.add(atmP);
+                }
+            }
+        }
+    }
+
+    private boolean checkAnno(
+            AnnotationMirror anno,
+            String canonicalTargetName,
+            Node n,
+            CheckController checkController,
+            boolean report) {
+        if (AnnotationUtils.hasElementValue(anno, "value")) {
+            UBQualifier.LessThanLengthOf qual =
+                    (UBQualifier.LessThanLengthOf) UBQualifier.createUBQualifier(anno);
+
+            List<String> strings = new ArrayList<>();
+            for (String s : qual.getArrays()) {
+                strings.add(s);
+            }
+            strings.addAll(qual.getOffsetsAsStrings());
+
+            List<String> canonicalStrings = new ArrayList<>();
+            for (String s : strings) {
+                try {
+                    FlowExpressions.Receiver r =
+                            analysis.getTypeFactory()
+                                    .getReceiverFromJavaExpressionString(
+                                            s, analysis.getTypeFactory().getPath(n.getTree()));
+                    canonicalStrings.add(r.toString());
+                    while (r instanceof FlowExpressions.FieldAccess) {
+                        // I included an exception here for "this", because otherwise the rules don't make sense - any
+                        // field, including final ones, would be invalidated, since it's impossible to know if this is
+                        // final or not. But this can't be modified by calling other code, so it's fine.
+                        if (checkController == CheckController.NONFINAL_REFS_AND_METHOD_CALLS
+                                || checkController
+                                        == CheckController.NONFINAL_REFS_AND_METHOD_CALLS_METHOD) {
+                            if (!((FlowExpressions.FieldAccess) r)
+                                            .getReceiver()
+                                            .toString()
+                                            .equals("this")
+                                    && !((FlowExpressions.FieldAccess) r).isFinal()) {
+                                if (report
+                                        && checkController
+                                                != CheckController
+                                                        .NONFINAL_REFS_AND_METHOD_CALLS_METHOD) {
```

Comment:
This whole method is unreadable.  Could you please do something to reduce the indentation level so that the code formats with less line breaks?  (Extract code to helper methods?  Only put code that might throw an exception in the try block?